Add some Events when containers die

kensimon opened the pull request:
**What this PR does / why we need it**:

Fix #38532

Ref #6014 #5927 #37616

The current events generated when containers die (or need to be killed, etc) give very minimal information when a container dies.  The first time it dies, you get nothing at all, and only if it enters a backoff interval do you get an actual message.  (And it looks very strange: `Error syncing pod, skipping`?)

This adds some explicit "container death" events, explaining what the exit code was at the time we noticed the container is no longer around.

Here is some sample output:

```
Events:
  FirstSeen	LastSeen	Count	From			SubObjectPath		Type		Reason		Message
  ---------	--------	-----	----			-------------		--------	------		-------
  27s		27s		1	default-scheduler				Normal		Scheduled	Successfully assigned dyer-1858799683-fr8kt to 127.0.0.1
  24s		24s		1	kubelet, 127.0.0.1	spec.containers{dyer}	Normal		Created		Created container with id b8787801533c4a7a2487592e728d4a4bb4de1e2cb40a6e5fcb8e89d498115609
  24s		24s		1	kubelet, 127.0.0.1	spec.containers{dyer}	Normal		Started		Started container with id b8787801533c4a7a2487592e728d4a4bb4de1e2cb40a6e5fcb8e89d498115609
  26s		2s		2	kubelet, 127.0.0.1	spec.containers{dyer}	Normal		Pulling		pulling image "gcr.io/heptio-images/dyer"
  2s		2s		1	kubelet, 127.0.0.1	spec.containers{dyer}	Warning		ExitFailure	Container "dyer" exited with status 1.
  24s		1s		2	kubelet, 127.0.0.1	spec.containers{dyer}	Normal		Pulled		Successfully pulled image "gcr.io/heptio-images/dyer"
  1s		1s		1	kubelet, 127.0.0.1	spec.containers{dyer}	Normal		Created		Created container with id 8f8fe866e878930c3a87afb129ef7e44fa61d193b53b0a0602d8edbff4957564
  1s		1s		1	kubelet, 127.0.0.1	spec.containers{dyer}	Normal		Started		Started container with id 8f8fe866e878930c3a87afb129ef7e44fa61d193b53b0a0602d8edbff4957564
```

**Special notes for your reviewer**:

- Container restart conditions are noticed in the `computePodContainerChanges` function, which certainly works, but I'm not 100% sure this is the best place to include them.  Advice here would be appreciated if I should do this in another place.

- Containers exiting with status 0 are "Normal" events, where nonzero is "Warning".  New events for containers becoming unhealthy and getting killed are Warnings. The existing sandbox change event has been changed to Warning.

Commit:
Add some Events when containers die

kensimon commented:
cc @timstclair

timothysc commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -444,21 +444,41 @@ func (m *kubeGenericRuntimeManager) computePodContainerChanges(pod *v1.Pod, podS
+			changes.ContainersToStart[index] = fmt.Sprintf("Starting container %+v.", container)
+			continue
+		} else if containerStatus.State != kubecontainer.ContainerStateRunning {
+			// TODO: is there a better place to report failed containers?
```
Ideally I would like to centralize these messages such that we could have ~=.rc file that could be internationalized. 

/cc @brendandburns

derekwaynecarr replied:
@timothysc -- i was not aware of the desire to i8n event messages.  is there something spec out in this space?

timothysc replied:
Centralizing event messages for debugging purposes mashed well with @brendandburns desire for internationalization.

timothysc commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -444,21 +444,41 @@ func (m *kubeGenericRuntimeManager) computePodContainerChanges(pod *v1.Pod, podS
 	changes.InitFailed = initFailed
 
+	podRef, err := ref.GetReference(api.Scheme, pod)
+	if err != nil {
```
can podRef be nil on error?

kensimon replied:
(This has since changed to containerRef to reference the specific container in the event instead of the pod, but...)

Yes, but Eventf tolerates this.  I'm not sure what the best error handling would be here, we certainly don't want to bail out but I definitely wouldn't want future developers using podRef in this scope assuming it's non-nil.

I saw this in a few other places in the codebase (not bailing out if a ref used for Events isn't found), so I carried the behavior forward.

derekwaynecarr commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -548,7 +571,7 @@ func (m *kubeGenericRuntimeManager) SyncPod(pod *v1.Pod, _ v1.PodStatus, podStat
 		}
 		if podContainerChanges.SandboxID != "" {
-			m.recorder.Eventf(ref, v1.EventTypeNormal, "SandboxChanged", "Pod sandbox changed, it will be killed and re-created.")
+			m.recorder.Eventf(ref, v1.EventTypeWarning, "SandboxChanged", "Pod sandbox changed, it will be killed and re-created.")
```
can we move SandboxChanged to the same location as other event declarations?

kensimon replied:
ack

kensimon commented:
Made some changes to smooth some things out a bit, addressing some of the comments.

Mainly, the event object is now the container itself rather than the pod, and the `Events` messages no longer have the blind verbosity that the log messages have.

The reason is that the use case for logs and for events are different: Events are already bound to a scoping object (like a container or pod), so there's no reason to re-include all that information in the event message.  Having less noise makes the event list easier to grasp.

Here's some sample output, after making a container unhealthy, letting it come back up, then killing its sandbox container:

```
Events:
  FirstSeen	LastSeen	Count	From			SubObjectPath			Type		Reason			Message
  ---------	--------	-----	----			-------------			--------	------			-------
  4m		4m		2	default-scheduler					Warning		FailedScheduling	no nodes available to schedule pods
  4m		4m		1	default-scheduler					Normal		Scheduled		Successfully assigned lifecycler-2022296878-9bvt4 to 127.0.0.1
  4m		4m		1	kubelet, 127.0.0.1	spec.containers{lifecycler}	Normal		Started			Started container with id 07b66751a8b37935666c345fe41787118e423ebefc2d73428a2ac333d456fc40
  4m		4m		1	kubelet, 127.0.0.1	spec.containers{lifecycler}	Normal		Created			Created container with id 07b66751a8b37935666c345fe41787118e423ebefc2d73428a2ac333d456fc40
  3m		3m		1	kubelet, 127.0.0.1	spec.containers{lifecycler}	Warning		Restarting		Container has become unhealthy, it will be killed and recreated
  3m		3m		1	kubelet, 127.0.0.1	spec.containers{lifecycler}	Normal		Killing			Killing container with id docker://07b66751a8b37935666c345fe41787118e423ebefc2d73428a2ac333d456fc40: Container is unhealthy
  3m		3m		1	kubelet, 127.0.0.1	spec.containers{lifecycler}	Normal		Created			Created container with id 01595279b366fe24bdaf13b710b5517682173bd73b9ef12ab7c12f62f8b42c5a
  3m		3m		1	kubelet, 127.0.0.1	spec.containers{lifecycler}	Normal		Started			Started container with id 01595279b366fe24bdaf13b710b5517682173bd73b9ef12ab7c12f62f8b42c5a
  31s		31s		1	kubelet, 127.0.0.1					Warning		SandboxChanged		Pod sandbox has changed, containers will be killed and recreated
  31s		31s		1	kubelet, 127.0.0.1	spec.containers{lifecycler}	Warning		ExitFailure		Container exited with status 137
  4m		30s		3	kubelet, 127.0.0.1	spec.containers{lifecycler}	Normal		Pulled			Successfully pulled image "gcr.io/heptio-images/lifecycler:latest"
  4m		30s		3	kubelet, 127.0.0.1	spec.containers{lifecycler}	Normal		Pulling			pulling image "gcr.io/heptio-images/lifecycler:latest"
  29s		29s		1	kubelet, 127.0.0.1	spec.containers{lifecycler}	Normal		Created			Created container with id 4cc940aa70269a58db6ac56f89abbd12e0bf5c5a6bd019529a9f5133c70e3415
  29s		29s		1	kubelet, 127.0.0.1	spec.containers{lifecycler}	Normal		Started			Started container with id 4cc940aa70269a58db6ac56f89abbd12e0bf5c5a6bd019529a9f5133c70e3415
```

feiskyer commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -447,19 +457,35 @@ func (m *kubeGenericRuntimeManager) computePodContainerChanges(pod *v1.Pod, podS
+			if containerStatus.ExitCode == 0 {
+				m.recorder.Eventf(containerRef, v1.EventTypeNormal, events.ExitSuccess, exitMessage)
+			} else {
+				m.recorder.Eventf(containerRef, v1.EventTypeWarning, events.ExitFailure, exitMessage)
```
Will this generate many repeated events for exited containers?

kensimon replied:
Looks like yes... didn't see this in testing until now.  I'll look into how to fix this.

kensimon replied:
Ok this is a little harder than I thought, since there's no "previous state" in scope at this point.

I'm exploring plumbing `existingStatus` in from kubelet.go:1389 through to SyncPod, which seems to work well.  I'll do some more testing and amend.

feiskyer commented:
@k8s-bot ok to test

kensimon commented:
Update: I'm now plumbing the current (existing) v1.PodStatus from the kubelet statusManager into the SyncPod function, using the second argument of SyncPod.  (See https://github.com/kensimon/kubernetes/blob/container-events-pt1/pkg/kubelet/kubelet.go#L1389)

Both the generic runtime manager and the rkt manager currently ignore the second param, (which is already a v1.PodStatus object), leaving me to wonder what the original intent was to plumb an API PodStatus through to the SyncPod function in the first place... (currently we send a newly-constructed v1.PodStatus object, which doesn't contain historical container status information because it's created on-the-fly.)

The point of this is to only emit the "Container exited" events when the container is dead but the previous status shows it as not dead.

I *think* this is necessary because all of the pod synchronization routines seem to be idempotent (or at least, they don't care about previous state), and there's no place that I can find to say "this is where the container flipped from healthy to exited" other than using the cached API object.  It's a little bit of a hack, so I'm wondering if anyone knows a better way.

kensimon commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -447,19 +457,42 @@ func (m *kubeGenericRuntimeManager) computePodContainerChanges(pod *v1.Pod, podS
+
+		var recordedContainerState v1.ContainerState
+		if len(existingStatus.ContainerStatuses) > index {
+			recordedContainerState = existingStatus.ContainerStatuses[index].State
```
Definitely need to fix this... I should be finding the container by name instead, but I got lost looking for the helper function for this that accepts a v1.PodStatus as opposed to a kubecontainer.PodStatus...

kensimon commented:
@Random-Liu @yifan-gu @feiskyer poke... review comments have been addressed, I'd appreciate some feedback on whether plumbing `existingStatus` into `SyncPod` is a good approach or not.

kensimon commented:
@yifan-gu  @Random-Liu any chance someone can take a look at this?

timothysc commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -434,6 +434,16 @@ func (m *kubeGenericRuntimeManager) computePodContainerChanges(pod *v1.Pod, podS
+	if err != nil {
+		glog.Errorf("Couldn't make a ref to pod %q: '%v'", format.Pod(pod), err)
+	}
+
```
We continue on here even though there could have been an error?

kensimon replied:
See comment.  Eventf handles ref being nil, and it's not a reason to stop going.  (This happens in several other places in the kubelet code, I'm just following convention)

timothysc commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -434,6 +434,16 @@ func (m *kubeGenericRuntimeManager) computePodContainerChanges(pod *v1.Pod, podS
+	}
+
+	if sandboxChanged && sandboxID != "" {
+		m.recorder.Eventf(podRef, v1.EventTypeWarning, events.SandboxChanged, "Pod sandbox has changed, containers will be killed and recreated")
```
so the phrase containers will be killed and recreated slightly confuses me b/c there are multiple levels of containers, which container is recreated?  The sandbox cgroup?

timothysc replied:
Is there any good in adding details about the pod in question?

kensimon replied:
All of them, and no, I don't want to add details about the pod because the event is already linked to the pod.  It "stutters" when the pod mentioned twice in the event, IMO.

timothysc commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -447,19 +457,46 @@ func (m *kubeGenericRuntimeManager) computePodContainerChanges(pod *v1.Pod, podS
+		if err != nil {
+			glog.Errorf("Couldn't make a ref to pod %q, container %v: '%v'", format.Pod(pod), container.Name, err)
+		}
+
```
Same comment about continuing after error.

timothysc commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -447,19 +457,46 @@ func (m *kubeGenericRuntimeManager) computePodContainerChanges(pod *v1.Pod, podS
+				// The container was running but currently is not. Record that it exited.
+				exitMessage := fmt.Sprintf("Container exited with status %d", containerStatus.ExitCode)
+				if containerStatus.ExitCode == 0 {
+					m.recorder.Eventf(containerRef, v1.EventTypeNormal, events.ExitSuccess, exitMessage)
```
Does this include pod details, same question below.

kensimon replied:
Same response as above, including not wanting to list the container (the event's InvolvedObject is already the container's name)

timothysc commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -447,19 +457,46 @@ func (m *kubeGenericRuntimeManager) computePodContainerChanges(pod *v1.Pod, podS
 	for index, container := range pod.Spec.Containers {
 		containerStatus := podStatus.FindContainerStatusByName(container.Name)
-		if containerStatus == nil || containerStatus.State != kubecontainer.ContainerStateRunning {
+		containerRef, err := kubecontainer.GenerateContainerRef(pod, &container)
```
Could the logic from here and below be put into a separate checking f(n)?

kensimon replied:
Not sure, maybe we can just discard the error and put the log message in the `GenerateContainerRef` function to begin with?  I'd be good with that.

timothysc commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -447,19 +457,46 @@ func (m *kubeGenericRuntimeManager) computePodContainerChanges(pod *v1.Pod, podS
-				message := fmt.Sprintf("Container %+v is dead, but RestartPolicy says that we should restart it.", container)
-				glog.Info(message)
-				changes.ContainersToStart[index] = message
+				glog.Infof("Container %+v is dead, but RestartPolicy says that we should restart it.", container)
```
Info?  or Warning?

kensimon replied:
I don't think it's a warning from kubelet's perspective, it's normal that kubelet restarts containers.  It's abnormal from the user's perspective (ie. the person who's running their app), but that's what events are for to begin with IMO.

timothysc commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -471,6 +508,8 @@ func (m *kubeGenericRuntimeManager) computePodContainerChanges(pod *v1.Pod, podS
 				message := fmt.Sprintf("Failed to initialize pod. %q will be restarted.", container.Name)
 				glog.V(1).Info(message)
+
+				m.recorder.Eventf(containerRef, v1.EventTypeWarning, events.ContainerRestarting, message)
```
Shouldn't the log and event both be warning?

kensimon replied:
See comment above, they're different audiences.  Kubelet is technically operating normally if it restarts dead pods.

derekwaynecarr replied:
it's possible i missed other areas, but I *think* this is prone to the same problem I fixed in https://github.com/kubernetes/kubernetes/pull/46246/files#diff-44f60dd6d99cb695e5f333647ebd0703R449

I worry we should be adapting these ;-)  i would feel better if we have one common function like generateContainerEvent that is called when good or bad so I know its right everywhere.

timothysc commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -479,10 +518,9 @@ func (m *kubeGenericRuntimeManager) computePodContainerChanges(pod *v1.Pod, podS
-			changes.ContainersToStart[index] = message
+			glog.Infof("Pod %q container %q hash changed (%d vs %d), it will be killed and recreated.", pod.Name, container.Name, containerStatus.Hash, expectedHash)
+			changes.ContainersToStart[index] = "Container hash has changed"
+			m.recorder.Eventf(containerRef, v1.EventTypeNormal, events.ContainerRestarting, "Container hash has changed, it will be killed and recreated")
```
This is normal?

kensimon replied:
Hmm, that one should probably be a warning.  Agreed.

timothysc commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -491,10 +529,13 @@ func (m *kubeGenericRuntimeManager) computePodContainerChanges(pod *v1.Pod, podS
+		if pod.Spec.RestartPolicy == v1.RestartPolicyNever {
+			m.recorder.Eventf(containerRef, v1.EventTypeWarning, events.ContainerRestarting, "Container has become unhealthy, and will be killed")
+		} else {
+			glog.Infof("pod %q container %q is unhealthy, it will be killed and recreated.", format.Pod(pod), container.Name)
```
s/Infof/Warningf

timothysc commented:
@Random-Liu @yujuhong PTAL.

kensimon commented:
Pushed a change to replace the one Normal event with a Warning (for when the container hash changes.)

In the rest of cases, my reasoning is this:

- The audience for Events and the audience for kubelet logs may be different... a container becoming unhealthy may be a Warning for the audience that's concerned with whether the container is working properly, but it's probably not a warning for kubelet itself, since kubelet is still behaving normally.

- Event messages don't need to contain repeated information from their InvolvedObject.  An event tied to a Pod doesn't need to have the pod name in the message, and an event tied to a Container doesn't need to have the Container name *or* the Pod name, since all of that information is in the event metadata.  So I've written the messages in a way is contextual to the object and doesn't repeat information.  (Note, this is not the case for kubelet logs, which don't have that information in scope, and so the log message needs to contain all the details.)

kensimon commented:
@k8s-bot pull-kubernetes-federation-e2e-gce test this

timothysc reviewed (approved):
/lgtm

timothysc commented:
@Random-Liu ping re: approval.

derekwaynecarr commented:
because i found myself debugging another production issue today related to event spam, this makes me nervous even though as user i like it.  if i have a restart always pod that starts and exits quickly, how many net new events will this cover?  if my docker container oom kills, will this result in any net new events?

timothysc commented:
@derekwaynecarr - we're working on server side aggregation btw.

kensimon commented:
@derekwaynecarr for scenarios where there's no issues initializing (ie. docker is working ok, the sandbox container is created normally), there should just be one new event per crash loop: what happened to the container.  (Did it exit successfully or not.)  The rest of the changes are to better explain what is going wrong with the initialization for edge cases like the sandbox not initializing, the hash changing, etc.

I think it's a worthwhile change to make, because no event currently tells you what happened to your container if it exits.

grodrigues3 commented:
@kensimon The 1.7 release team wants all PRs in the milestone to reference the github issue it addresses.  Can you reference the issue in the PR body please?  If there is not an issue, please file one.

@timothysc you're /lgtm never went through (though it looks like you attempted to do it prior to code freeze).  You mind trying again?

derekwaynecarr commented:
I want to understand what happens for following:
* a pod with restart policy = always, but whose container exits frequently
* a pod that induces an oom kill local to container, whose restart policy is always

assuming this doesn't induce too much net new traffic (and normal things like backoff kick in), this is fine for me.

kensimon commented:
Here's sample `kubectl describe pod` output for those scenarios:

https://gist.github.com/kensimon/1376c5016d384e8a6ac76075dd4f1a2f

timothysc commented:
@k8s-bot pull-kubernetes-e2e-gce-etcd3 test this

kensimon commented:
Sorry, but I had to do one quick fix to the behavior:

When the container dies, instead of checking if the previous status was "Running" before we exited, it now checks if it was "Not Terminated", which catches cases where the container died too quickly to notice that it was running.

Also, I dropped the Pod-Wide "sandbox changed" event, because it's redundant (you get the event for the containers that are restarted anyway), and it turns out the event was firing when a Pod is complete (ie. exits 0 on restartPolicy=OnFailure), which led to a very misleading event.

kensimon commented:
@k8s-ci-robot pull-kubernetes-e2e-gce-etcd3 test this

kensimon commented:
@k8s-ci-robot pull-kubernetes-verify test this

kensimon commented:
@k8s-bot pull-kubernetes-verify test this

kensimon commented:
@k8s-bot pull-kubernetes-e2e-gce-etcd3 test this

kensimon commented:
@k8s-bot pull-kubernetes-e2e-gce-etcd3 test this

timothysc commented:
@derekwaynecarr ping.

derekwaynecarr commented on pkg/kubelet/kubelet.go:
```diff
@@ -1607,7 +1607,7 @@ func (kl *Kubelet) syncPod(o syncPodOptions) error {
 
 	// Call the container runtime's SyncPod callback
-	result := kl.containerRuntime.SyncPod(pod, apiPodStatus, podStatus, pullSecrets, kl.backOff)
+	result := kl.containerRuntime.SyncPod(pod, existingStatus, podStatus, pullSecrets, kl.backOff)
```
why did you change this?

kensimon replied:
See my comments here: https://github.com/kubernetes/kubernetes/pull/45682#issuecomment-301137582

The SyncPod function was ignoring the second argument, I'm not sure why apiPodStatus was being sent to begin with.  (apiPodStatus appears to be the status object that will be sent back to the API server to report state... existingStatus is the status that the kubelet knew about from the last time it checked.)

Since we need the previous status plumbed through, and SyncPod wasn't using the parameter before, I re-used that positional parameter rather than adding a new one.

derekwaynecarr replied:
add a comment why you use existingStatus in the code here instead of apiPodStatus.  i find existing status versus apiPodStatus confusing as well.  one should be previousStatus the other should be observedStatus, can we clean that up?

derekwaynecarr commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -542,20 +573,12 @@ func (m *kubeGenericRuntimeManager) computePodContainerChanges(pod *v1.Pod, podS
 //  5. Create init containers.
 //  6. Create normal containers.
-func (m *kubeGenericRuntimeManager) SyncPod(pod *v1.Pod, _ v1.PodStatus, podStatus *kubecontainer.PodStatus, pullSecrets []v1.Secret, backOff *flowcontrol.Backoff) (result kubecontainer.PodSyncResult) {
+func (m *kubeGenericRuntimeManager) SyncPod(pod *v1.Pod, existingStatus v1.PodStatus, podStatus *kubecontainer.PodStatus, pullSecrets []v1.Secret, backOff *flowcontrol.Backoff) (result kubecontainer.PodSyncResult) {
```
can you update the signature in runtime.go for SyncPod for what you expect apiPodStatus to hold?

kensimon replied:
ack

derekwaynecarr commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -424,7 +422,7 @@ func checkAndKeepInitContainers(pod *v1.Pod, podStatus *kubecontainer.PodStatus,
 
 // computePodContainerChanges checks whether the pod spec has changed and returns the changes if true.
-func (m *kubeGenericRuntimeManager) computePodContainerChanges(pod *v1.Pod, podStatus *kubecontainer.PodStatus) podContainerSpecChanges {
+func (m *kubeGenericRuntimeManager) computePodContainerChanges(pod *v1.Pod, podStatus *kubecontainer.PodStatus, existingStatus *v1.PodStatus) podContainerSpecChanges {
```
this is confusing as well.

existingStatus == prevObservedStatus
podStatus == observedStatus

derekwaynecarr reviewed (changes requested):
a few naming requests

and a concern that we should be going through events.ToObjectReference for some of the failure scenarios.  i would feel better if all the event stuff that dealt with object references were in a common function.

kensimon commented:
@derekwaynecarr PTAL

- Renamed the arguments to better reflect what their purpose is (including in the interface definition for SyncPod)
- (per our Slack convo) refactored the SyncPod function to move the `prevObservedStatus` to the last flag, since it's confusing having it be the 2nd argument if it's of auxiliary importance.

kensimon commented:
@k8s-bot pull-kubernetes-e2e-gce-etcd3 test this

derekwaynecarr commented:
@kensimon -- i am not ignoring this.  i spent the better part of today debugging pathological users on a cluster that had pods with always failing readiness probes, or pods that never start.  one of the things i am concerned about is that all of the `FailedSync` events have a unique containerId entries as well.

/cc @eparis -- i need to reconcile the bad offending pods in the one cluster we have to see if this would have made things even worse.

see related attempt to add spam filtering: https://github.com/kubernetes/kubernetes/pull/47367

dchen1107 commented:
Move this pr to v1.8 per above discussion. Today we have event spamming issue already. Unless we have a programatical way to resolve the issue, we don't think we should get the situation worse. Thanks, and sorry for the inconvenience.

timothysc commented:
imo this should probably go in now and @gmarek is working on fixing the degenerative spamming case.

gmarek commented:
I'm skeptical about this PR. I may be mistaken, but I believe that Container reference contains containerID, which changes with each instantiation, hence if we keep populating it, it won't be aggregated. This may cause serious problems @derekwaynecarr

DreadPirateShawn commented:
Is there any update on this? Reference issue #38532 is waiting for this or some alternative solution.

The key open question from that ticket, I think, is:

> The pod lifetime doc seems to read as if any actions that result in a container death should result in an event -- container exit success, container exit failure, container OOM, disk death -- if there are clear ways for a container to die unexpectedly and *not* generate an event, that might be worth calling out more clearly? (I know it's easy to overthink docs -- just explaining our reasoning when filing.)

timothysc commented:
Abandoning this PR.
